Add MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS to limit concurrent scorer execution

debu-sinha opened the pull request:
### Related Issues/PRs

Resolves #18212

### What changes are proposed in this pull request?

Adds a new environment variable `MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS` to control the maximum number of concurrent scorer workers during evaluation. This helps prevent rate limiting errors when using external LLM APIs as judges.

**Root Cause:** The scorer execution loop in `harness.py` uses `max_workers=len(scorers)` which is unbounded. Combined with the outer loop (`MLFLOW_GENAI_EVAL_MAX_WORKERS=10`), this can result in `10 × num_scorers` concurrent API calls - easily hitting rate limits with services like Gemini free tier.

**Changes:**
- Add `MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS` environment variable (default: 10)
- Modify scorer execution to use `min(len(scorers), MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS.get())`
- Add test for the new environment variable
- Add documentation to the GenAI Evaluation quickstart guide

**Usage:**
```bash
# Limit concurrent scorers to 2
export MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS=2

# Run scorers sequentially (eliminates rate limiting entirely)
export MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS=1

# Combined control: 5 items × 2 scorers = max 10 concurrent API calls
export MLFLOW_GENAI_EVAL_MAX_WORKERS=5
export MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS=2
```

### How is this PR tested?

- [x] New unit/integration tests

Added `test_max_scorer_workers_env_var` test to verify the environment variable correctly limits concurrent scorer execution.

### Does this PR require documentation update?

- [x] Yes. I've updated:
  - [x] Instructions

Added a tip section to the GenAI Evaluation quickstart guide documenting both `MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS` and `MLFLOW_GENAI_EVAL_MAX_WORKERS` environment variables.

### Release Notes

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added `MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS` environment variable to control concurrent scorer execution during `mlflow.genai.evaluate()`, helping prevent rate limiting errors when using external LLM APIs as judges.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [x] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

debu-sinha committed:
Add MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS to limit concurrent scorer execution

Adds a new environment variable to control the maximum number of concurrent
scorer workers during evaluation. This helps prevent rate limiting errors
when using external LLM APIs as judges.

Changes:
- Add MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS env var (default: 10)
- Modify harness.py to use min(num_scorers, max_scorer_workers)
- Add test for the new environment variable

The total concurrent API calls is now bounded by:
MLFLOW_GENAI_EVAL_MAX_WORKERS × MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS

Users can set MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS=1 to run scorers
sequentially, which eliminates rate limiting issues entirely.

Addresses #18212

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

BenWilson2 commented:
The approach looks solid - can we add this as a docs note for the scorers documentation? Having mention of this in the quickstart https://mlflow.org/docs/latest/genai/eval-monitor/quickstart/ is probably a good place as this is likely something that users encounter frequently when getting started.

debu-sinha committed:
Add rate limiting documentation to GenAI evaluation quickstart

Document MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS and MLFLOW_GENAI_EVAL_MAX_WORKERS
environment variables in the quickstart guide to help users avoid rate limiting
errors when using LLM-as-a-Judge scorers with external API providers.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented:
Thanks for the feedback! I've added a tip section to the quickstart guide documenting both `MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS` and `MLFLOW_GENAI_EVAL_MAX_WORKERS` environment variables, including how they interact and examples for avoiding rate limiting.

See the updated docs preview: https://pr-19248--mlflow-docs-preview.netlify.app/docs/latest/genai/eval-monitor/quickstart/

debu-sinha commented:
Hi @BenWilson2 - I've added the documentation to the quickstart guide as you suggested. The preview is available here: https://pr-19248--mlflow-docs-preview.netlify.app/docs/latest/genai/eval-monitor/quickstart/

Let me know if there's anything else you'd like me to adjust!

B-Step62 commented on mlflow/environment_variables.py:
```diff
@@ -674,6 +674,15 @@ def get(self):
 #: (default: ``10``)
 MLFLOW_GENAI_EVAL_MAX_WORKERS = _EnvironmentVariable("MLFLOW_GENAI_EVAL_MAX_WORKERS", int, 10)
 
+#: Maximum number of concurrent scorer workers to use when running multiple scorers
```
nit: Let's mention that the worker size will not be larger than the number of scorers.

B-Step62 reviewed (approved):
Thanks for the enhancement @debu-sinha! The change looks great, left a few nits on doc update.

debu-sinha committed:
Move rate limiting docs to FAQ and update env var description

Address reviewer feedback:
- Move detailed concurrency documentation from quickstart.mdx to faq.mdx
- Add scorer-level concurrency section explaining MLFLOW_GENAI_EVAL_MAX_SCORER_WORKERS
- Update environment_variables.py to mention worker count won't exceed number of scorers
- Clarify total concurrent API calls formula in FAQ

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

B-Step62 committed:
Merge branch 'master' into feature/rate-limit-evaluation

debu-sinha committed:
Merge branch 'master' into feature/rate-limit-evaluation
